### PR TITLE
Redesign the PackageJsonLookup API

### PR DESCRIPTION
**Pull request by pgonzal**
This is a breaking change for PackageJsonLookup, which clarifies the API names, and expands the cache to store individual package.json fields beyond just the package name.

Rush's IPackageJson interface is promoted to be part of **node-core-library**.

### CODE REVIEW

File: libraries/node-core-library/src/PackageJsonLookup.ts [hunk 1]
```diff
@@ -40,56 +57,152 @@ export class PackageJsonLookup {
...
+      if (!loadedPackageJson.name) {
+        throw new Error(`Error reading "${jsonFilename}":\n  `
+          + 'The required field "name" was not found');
+      }
+      if (!loadedPackageJson.version) {
+        throw new Error(`Error reading "${jsonFilename}":\n  `
+          + 'The required field "version" was not found');
+      }
+
+      if (this._loadExtraFields) {
```
Comment (iclanton) [inline libraries/node-core-library/src/PackageJsonLookup.ts]: >_loadExtraFields [](start = 15, length = 16)

What about an optional parameter that lets you list out properties that you want?

Comment (pgonzal) [reply libraries/node-core-library/src/PackageJsonLookup.ts]: It's a good idea, but I'd like to wait until someone really needs it before designing it.

Comment (pgonzal) [reply libraries/node-core-library/src/PackageJsonLookup.ts]: It's a good idea, but I'd like to wait until someone really needs it before designing it.

---
In reply to: [175201618](https://github.com/Microsoft/web-build-tools/pull/564#discussion_r175201618) [](ancestors = 175201618)

**Comment by pgonzal** [comment]
Skipped Travis build because Travis is experiencing an outage

**Pull request merged**